### PR TITLE
Refactor functional testing to improve resiliency and simplify addition of new tests

### PR DESCRIPTION
**Pull request by paulbreuler**
Refactoring monolithic component functional testing to remove dependency on Gulp and improve overall functionality. Proposing change for N+1 release with improvements to existing test occurring in parallel to ensure coverage and quality in existing code. 

### Intent
- Simplify Mocha test structure allowing new tests to be added with minimal overhead.
- Improve logging
- Create a better end to end dev inner-loop experience, improving local testing and debugging

### Implementation notes:
Verbose output is available simply by setting the following environment variables for a given pipeline run.
```
DEBUG = taskTest:*
DEBUG_COLORS = 1 
```

This is set already for local testing via VS Code launch configuration.

### CODE REVIEW

File: test/functional-test/functional-test-lib/taskTestRunner.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    this.taskDirectory = taskDirectory;
+  }
+
+  runTask(): TaskResult{
+    const normalizedTaskPath = this.normalizeAbsoluteTaskPath()
+    this.taskResult = cp.spawnSync('node', [normalizedTaskPath], { encoding: 'utf-8', cwd: this.taskDirectory });
+    console.debug(this.taskResult.stdout);
+    this.validateTaskRun();
+    var envVar = this.setOutputEnvironmentVariables();
+    return {processResult: this.taskResult, outputEnvironmentVariable: envVar };
```
Comment (tehcrashxor) [inline test/functional-test/functional-test-lib/taskTestRunner.ts:37]: Style nit: space after the opening brace
`return { processResult`
and perhaps some linting rules to enforce this, here or in another PR?

File: test/functional-test/functional-test-lib/taskTestRunner.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  private normalizeAbsoluteTaskPath() {
+    return (path.join(this.taskDirectory, this.taskInfo.path)).replace(/\\/g, '/');
+  }
+
+  private setOutputEnvironmentVariables(): EnvironmentVariableDefinition | undefined {
+    if(!this.taskResult) return;
```
Comment (tehcrashxor) [inline test/functional-test/functional-test-lib/taskTestRunner.ts:45]: Style nit: space between if and parenthesis
`if (!this.taskResult) return;`
ditto below in line 55

File: test/functional-test/functional-test-lib/taskTestRunner.ts (hunk 1 above)
Comment (paulbreuler) [reply test/functional-test/functional-test-lib/taskTestRunner.ts:37]: Agreed, let's add some auto linting if possible, to ensure a general pattern / style is followed.

**Review by tehcrashxor — APPROVED** [review]

File: test/functional-test/taskTestInput.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    {
+      name: 'export-solution',
+      path: '/tasks/export-solution/export-solution-v0',
+      inputVariables: [
+        { name: 'solutionName', value: 'emptySolution' },
+        { name: 'SolutionVersionNumber', value: '0.42.0.2' },
+        { name: 'SolutionOutputFile', value: path.join(solutionTestOutputRootDirectory, 'exported-solution', `solution_${new Date().toJSON().slice(0,10)}.zip`) },
+      ]
+    },
+    // {
```
[the remark this answers is not in the document]
Comment (paulbreuler) [reply test/functional-test/taskTestInput.ts:99]: Correct, the current issue is that there is no code-based way to hide these. I've explored various solutions but landed on commenting out for now until a better solution is found.

File: test/functional-test/functional-test-lib/taskTestBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    createReadStream(packageToTest)
+      .pipe(unzip.Extract({ path: this.taskRootPath }))
+      .on("close", callBack.bind(this))
+      .on("error", (error: any) => {
+        throw new Error(`Failed to extract ${packageToTest} to ${this.taskRootPath}: error: ${error}`);
+      });
+  }
+
+  private resolvePackageToTestPath() {
+
```
[the remark this answers is not in the document]
Comment (paulbreuler) [reply test/functional-test/functional-test-lib/taskTestBuilder.ts]: Hopefully corrected when I updated the readability flow

File: test/functional-test/taskTestInput.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulbreuler) [reply test/functional-test/taskTestInput.ts:99]: To test them we need to 1) have the correct PAC CLI version and 2) have the feature flags set to on or use a daily build where all flags are on.

**Comment by paulbreuler** [comment]
Something to note: The code does have debug output to make it easier to track at a more verbose level as needed. 

To enable it we simply need to set an environment variable to ensure output is always visible.

  DEBUG = taskTest:*
  DEBUG_COLORS = 1
  
 Colors make it far easier to identify from which module the logs are generated. I did not want to modify the PR workflow in this branch and PR to keep the scope down to just adding the new test setup.

**Pull request merged**